dipy/reconst/multi_voxel.py

```diff
@@ -71,17 +100,58 @@ def new_fit(self, data, mask=None, **kwargs):
                 data_to_fit[ii : ii + vox_per_chunk]
                 for ii in range(0, data_to_fit.shape[0], vox_per_chunk)
             ]
-            fit_array[np.where(mask)] = np.concatenate(
-                (
-                    paramap(
-                        _parallel_fit_worker,
-                        chunks,
-                        func_args=[single_voxel_with_self],
-                        **kwargs,
-                    )
+            # FIXME: weights are now broken, DUE TO PULL REQUEST MADE WHILE MINE WAS WAITING
+            #        I was passing weights as a kwarg to make it work,
+            #        I don't think it will be possible to do in the way they have set-up here
+            #        thanks, DiPy team, you really do go above and beyond to make things difficult
+            print("in multi_voxel here are the keywords")
+            print(kwargs)
+            print(weights_is_array)
```

Comment:
Sorry - I am having a really hard to to understand this PR. Are these weights something that can be passed to the `__init__` method of the model instead of the `fit` method? As you can see in #3286, passing kwargs to the model fit is potentially quite brittle in the presence of parallelization.

Reply:
No the weights can't be passed to init, they must be passed to fit. Because for iteratively reweighed schemes, the weights are calculated repeatedly. They are passed as keywords because there was no other solution that was as flexible.

In summary, the PR allows for weights to be inputs to some for routines. There are functions to calculate these weights. Hence, it allowed a more generalizable way to do IRLS which wasn't in DiPy before.

Better testing of parallelization would check that keywords are passed properly.

Reply:
I think that what happened here is that both PRs introduced use of the keyword arguments to the 'fit' method that was not generally possible before. I think that a better solution than the one I am proposing in #3086 (which still seems very brittle to me) is to set up the parallelization keyword arguments in the init, leaving a clear path for this kind of use of the fit method keyword args. Does anyone see any issues with this? I do think that we were trying to avoid keyword args to 'fit' in the past to provide a uniform api across models but this PR demonstrates perhaps where we could deviate from that.

Reply:
I want to say that I considered many different options and determined that this was the best approach. Worth saying that, until recently, the tests were passing, and I extensively tested thanks to advice from DiPy team.

If I can do more to make clear what is going on here, I will, but I think my original PR message explains it fairly well: I generalize a few functions to take weights as keyword arguments (other than booleans), added some functions to calculate these weights from fit residuals, and added functions to do an iterative fit. I think the testing demonstrates that, except for the new parallelization issue in which keywords don't appear to get passed correctly, everything works. I even added some useful tests to help show it's working as expected.